### PR TITLE
Refactor proxy-set-headers on Ingress

### PR DESCRIPTION
**Pull request by pdabelf5**
This pull request significantly refactors and improves the handling of the `nginx.org/proxy-set-headers` annotation in the NGINX Ingress Controller codebase. The changes introduce a new, more robust parsing function for proxy headers, add support for merging headers from master and minion ingresses (with clear precedence), and update both the template rendering and tests to match the new logic. This results in more predictable, flexible, and testable behavior for custom proxy headers.

This change is a precursor to the work for External Auth which will make use of the Ingress `Location.ProxySetHeaders` struct.

**Proxy-Set-Headers Handling Improvements:**

* Introduced a new constant `ProxySetHeadersAnnotation` for `"nginx.org/proxy-set-headers"` to ensure consistent usage across the codebase.
* Replaced the old slice-based parsing of proxy-set-headers with a new function `ParseProxySetHeaders`, which parses a comma-separated string, trims whitespace, derives default NGINX `$http_` variable values when no value is provided, and skips invalid entries.
* Added a `MergeProxySetHeaders` function that merges master and minion annotations, giving priority to minion headers when names overlap. [[1]](diffhunk://#diff-892e8a3990127f43813b9de650b0e326202b6e2449fb1c886f28ae992521e205L299-R350) [[2]](diffhunk://#diff-92c0ea851013e748f27d1c7d7e4c1a9eda68c72fbe819f43f8feb286700862abR846-R851)
* Updated the annotation parsing logic to use the new constant and parsing function, ensuring correct extraction and processing of the annotation.
* Modified mergeable ingress configuration logic to use `MergeProxySetHeaders`, ensuring correct header precedence between master and minion ingresses.

**Template and Rendering Updates:**

* Updated the template logic to render each parsed proxy header using a loop, instead of relying on a helper that generated the entire block, making the template more transparent and maintainable.

**Testing Enhancements:**

* Rewrote and expanded tests for proxy-set-headers parsing and merging, covering edge cases such as whitespace, empty values, default value derivation, multiple headers, and correct precedence in merging.
* Updated test snapshots to match the new template output, ensuring tests reflect the new rendering logic. [[1]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL893) [[2]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL938) [[3]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL983) [[4]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1028) [[5]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1073) [[6]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1118) [[7]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1162) [[8]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1212) [[9]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1262) [[10]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1356) [[11]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1429) [[12]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1494) [[13]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1543) [[14]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1621) [[15]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1699) [[16]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1777) [[17]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1855) [[18]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1927) [[19]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL1955) [[20]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL2037) [[21]](diffhunk://#diff-0c1edb7ca2c7c99b5b3e3fee1e2d26c6ff2367981fb82093e27fd2ce42b6865eL2068)

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginx/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [ ] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

File: internal/configs/parsing_helpers.go [hunk 1]
```diff
@@ -296,20 +296,58 @@ func ParseProxyBuffersSpec(s string) (string, error) {
...
+	seen := make(map[string]bool)
+	var merged []version2.Header
+
+	for _, h := range minionHeaders {
+		seen[h.Name] = true
+		merged = append(merged, h)
+	}
+
+	for _, h := range masterHeaders {
+		if !seen[h.Name] {
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/parsing_helpers.go]: Addressed in upcoming commit

File: internal/configs/ingress.go [hunk 2]
```diff
@@ -843,6 +843,12 @@ func generateNginxCfgForMergeableIngresses(ncp NginxCfgParams) (version1.Ingress
 					loc.CORSEnabled = true
 				}
 				loc.MinionIngress = &minionNginxCfg.Ingress
+				// Merge proxy-set-headers: minion headers take priority over master headers.
+				masterAnnotation := ncp.mergeableIngs.Master.Ingress.Annotations["nginx.org/proxy-set-headers"]
+				minionAnnotation := minion.Ingress.Annotations["nginx.org/proxy-set-headers"]
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/ingress.go]: Addressed in an upcoming commit

File: internal/configs/version1/template_test.go [hunk 3]
```diff
@@ -4676,7 +4678,55 @@ var (
 	}
 )
 
+// parseTestProxySetHeaders is a test-only copy of the parsing logic from
+// configs.ParseProxySetHeaders, used to populate Location.ProxySetHeaders
+// in template tests without creating a circular import.
+func parseTestProxySetHeaders(annotation string) []version2.Header {
+	var headers []version2.Header
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/version1/template_test.go]: Moved parse & merge logic to version1 package

File: internal/k8s/validation.go [hunk 4]
```diff
@@ -558,41 +559,33 @@ func validateHTTPHeadersAnnotation(context *annotationValidationContext) field.E
...
-			allErrs = append(allErrs, field.Invalid(context.fieldPath, header, "invalid header syntax - spaces are not allowed in header"))
-		}
-
-		if len(parts) != 2 {
-			parts = append(parts, "")
+	for _, entry := range strings.Split(context.value, commaDelimiter) {
+		entry = strings.TrimSpace(entry)
+		if entry == "" {
+			continue
 		}
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/k8s/validation.go:574]: Addressed in upcoming commit

File: internal/configs/version1/nginx-plus.ingress.tmpl [hunk 5]
```diff
@@ -346,8 +346,9 @@ server {
 		{{- if $location.ClientBodyBufferSize }}
 		client_body_buffer_size {{$location.ClientBodyBufferSize}};
 		{{- end}}
-		{{- $proxySetHeaders := generateProxySetHeaders $location $.Ingress.Annotations }}
-		{{$proxySetHeaders}}
+		{{- range $header := $location.ProxySetHeaders}}
+		proxy_set_header {{ $header.Name }} "{{ $header.Value }}";
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/version1/nginx-plus.ingress.tmpl]: Added escaped string checking at header value validation step

File: internal/configs/version1/template_executor_test.go [hunk 6]
```diff
@@ -651,8 +651,9 @@ server {
 		proxy_read_timeout {{$location.ProxyReadTimeout}};
 		proxy_send_timeout {{$location.ProxySendTimeout}};
 		client_max_body_size {{$location.ClientMaxBodySize}};
-		{{- $proxySetHeaders := generateProxySetHeaders $location $.Ingress.Annotations -}}
-		{{$proxySetHeaders}}
+		{{- range $header := $location.ProxySetHeaders}}
+		proxy_set_header {{ $header.Name }} "{{ $header.Value }}";
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/version1/template_executor_test.go]: Added escaped string checking at header value validation step

File: internal/configs/configurator_test.go [hunk 7]
```diff
@@ -2575,8 +2575,9 @@ server {
 		proxy_read_timeout {{$location.ProxyReadTimeout}};
 		proxy_send_timeout {{$location.ProxySendTimeout}};
 		client_max_body_size {{$location.ClientMaxBodySize}};
-		{{- $proxySetHeaders := generateProxySetHeaders $location $.Ingress.Annotations -}}
-		{{$proxySetHeaders}}
+		{{- range $header := $location.ProxySetHeaders}}
+		proxy_set_header {{ $header.Name }} "{{ $header.Value }}";
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/configurator_test.go]: Added escaped string checking at header value validation step

File: internal/configs/version1/nginx.ingress.tmpl [hunk 8]
```diff
@@ -261,8 +261,9 @@ server {
 		{{- if $location.ClientBodyBufferSize }}
 		client_body_buffer_size {{$location.ClientBodyBufferSize}};
 		{{- end}}
-		{{- $proxySetHeaders := generateProxySetHeaders $location $.Ingress.Annotations -}}
-		{{$proxySetHeaders}}
+		{{- range $header := $location.ProxySetHeaders}}
+		proxy_set_header {{ $header.Name }} "{{ $header.Value }}";
```
[the remark this answers is not in the document]
Comment (pdabelf5) [reply internal/configs/version1/nginx.ingress.tmpl]: Added escaped string checking at header value validation step

File: internal/configs/version1/template_test.go [hunk 9]
```diff
@@ -4677,6 +4679,10 @@ var (
 )
 
 func createProxySetHeaderIngressConfig(masterAnnotations map[string]string, coffeeAnnotations map[string]string, teamAnnotations map[string]string) IngressNginxConfig {
+	masterPSH := masterAnnotations["nginx.org/proxy-set-headers"]
+	coffeePSH := coffeeAnnotations["nginx.org/proxy-set-headers"]
+	teaPSH := teamAnnotations["nginx.org/proxy-set-headers"]
```
Comment (sean-breen) [inline internal/configs/version1/template_test.go]: typo, `teaAnnotations`

Comment (pdabelf5) [reply internal/configs/version1/template_test.go]: Thanks @sean-breen, that should be fixed now

File: internal/configs/version1/template_executor_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (javorszky) [reply internal/configs/version1/template_executor_test.go]: The rest of the templates use printf %q, is this something we should revisit here, even though it's just a test?

File: internal/configs/version1/template_helper.go [hunk 10]
```diff
@@ -273,19 +155,18 @@ func extractOriginalPath(processedPath string) string {
...
-	"hasPrefix":               strings.HasPrefix,
-	"hasSuffix":               strings.HasSuffix,
-	"toLower":                 strings.ToLower,
-	"toUpper":                 strings.ToUpper,
-	"replaceAll":              strings.ReplaceAll,
-	"makeLocationPath":        makeLocationPath,
-	"makeRewritePattern":      makeRewritePattern,
-	"makeSecretPath":          commonhelpers.MakeSecretPath,
-	"makeOnOffFromBool":       commonhelpers.MakeOnOffFromBool,
-	"generateProxySetHeaders": generateProxySetHeaders,
```
Comment (javorszky) [inline internal/configs/version1/template_helper.go:288]: the entire change is reformatting because the `generateProxySetHeaders` was removed 🙂

File: internal/configs/version1/template_test.go [hunk 11]
```diff
@@ -4675,7 +4677,11 @@ var (
 	}
 )
 
-func createProxySetHeaderIngressConfig(masterAnnotations map[string]string, coffeeAnnotations map[string]string, teamAnnotations map[string]string) IngressNginxConfig {
+func createProxySetHeaderIngressConfig(masterAnnotations map[string]string, coffeeAnnotations map[string]string, teaAnnotations map[string]string) IngressNginxConfig {
```
Comment (javorszky) [inline internal/configs/version1/template_test.go]: The tiniest of nits, not important, but looks nicer like this 🙂 

```suggestion
func createProxySetHeaderIngressConfig(masterAnnotations, coffeeAnnotations, teaAnnotations map[string]string) IngressNginxConfig {
```

File: internal/configs/configurator_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (javorszky) [reply internal/configs/configurator_test.go]: PR uses printf %q elsewhere, should this be adapted to that style as well?

File: internal/configs/ingress.go [hunk 12]
```diff
@@ -841,6 +841,12 @@ func generateNginxCfgForMergeableIngresses(ncp NginxCfgParams) (version1.Ingress
 					loc.CORSEnabled = true
 				}
 				loc.MinionIngress = &minionNginxCfg.Ingress
+				// Merge proxy-set-headers: minion headers take priority over master headers.
+				masterAnnotation := ncp.mergeableIngs.Master.Ingress.Annotations[ProxySetHeadersAnnotation]
+				minionAnnotation := minion.Ingress.Annotations[ProxySetHeadersAnnotation]
+				if masterAnnotation != "" || minionAnnotation != "" {
```
Comment (javorszky) [inline internal/configs/ingress.go]: I think this check is not needed. The MergeProxySetHeaders calls ParseProxyHeaders, and both of those functions handle empty-empty inputs. There's a test case for MergeProxySetHeaders getting empty-empty. The result would be `nil`.

This is used in the template rendering in a rage block, which is perfectly happy to range over a nil `[]version2.Header` over an empty `[]version2.Header{}`.

File: internal/k8s/validation.go [hunk 13]
```diff
@@ -558,41 +559,37 @@ func validateHTTPHeadersAnnotation(context *annotationValidationContext) field.E
...
-		header = strings.TrimSpace(header)
 
-		parts := strings.SplitN(header, ":", 2)
-		if len(parts) == 1 && strings.Contains(header, " ") {
-			allErrs = append(allErrs, field.Invalid(context.fieldPath, header, "invalid header syntax - spaces are not allowed in header"))
-		}
-
-		if len(parts) != 2 {
-			parts = append(parts, "")
+	for _, entry := range strings.Split(context.value, commaDelimiter) {
```
Comment (javorszky) [inline internal/k8s/validation.go:569]: this is partially reimplementing the `version1.ParseProxySetHeaders` function. We could do this:

```golang
headers := version1.ParseProxySetHeaders(context.value)

for _, header := range headers {
    // validate if http header name
    // check if contains $
}
```

If we move the helpers to version2, we can remove a version1 -> version2 requirement, and all places that these are used also import version2 already anyways.

File: internal/k8s/validation_test.go [hunk 14]
```diff
@@ -1268,7 +1268,7 @@ func TestValidateNginxIngressAnnotations(t *testing.T) {
 
 		{
 			annotations: map[string]string{
-				"nginx.org/proxy-set-headers": "header-1",
```
Comment (javorszky) [inline internal/k8s/validation_test.go:1271]: tests should keep using the naked string rather than the constant.

In that case if the constant is changed for whatever reason, these would not automatically change and pass silently, which means a change to the constant would necessarily mean a change to the tests which signals intention rather than accidents.

File: internal/configs/version1/template_helper.go (hunk 10 above)
Comment (AlexFenlon) [reply internal/configs/version1/template_helper.go:288]: correct! generateProxySetHeaders is no longer needed as the logic is now in better places rather than right next to the template.

File: internal/configs/ingress.go (hunk 12 above)
Comment (javorszky) [reply internal/configs/ingress.go]: I meant "range block", not rage block 😡 😂

File: internal/k8s/validation.go (hunk 13 above)
Comment (pdabelf5) [reply internal/k8s/validation.go:569]: Is looked into this, if we re-used `ParseProxySetHeaders()` we would loose the warning if a header is missing.  For example "header1,,header3" should error regarding missing header

File: internal/k8s/validation_test.go (hunk 14 above)
Comment (pdabelf5) [reply internal/k8s/validation_test.go:1271]: Having the annotation as a string makes the test fail in way that makes it hard to debug if you change the const.  For example:
Results for string
```
 /Users/p.abel/git/github/nginx/kubernetes-ingress/internal/k8s/validation_test.go:4232: validateIngressAnnotations() returned 
 
            but expected 
            annotations.nginx.org/proxy-set-headers: Invalid value: "$header1": a valid HTTP header must consist of alphanumeric characters or '-' (e.g. 'X-Header-Name', regex used for validation is '[-A-Za-z0-9]+') 
            for the case of invalid nginx.org/proxy-set-headers annotation, single-value containing '$'
    --- FAIL: TestValidateNginxIngressAnnotations/invalid_nginx.org/proxy-set-headers_annotation,_single-value_containing_'{' 
```

Results for const
```
 /Users/p.abel/git/github/nginx/kubernetes-ingress/internal/k8s/validation_test.go:4232: validateIngressAnnotations() returned 
            annotations.nginx.org/proxy-set-headersdddd: Invalid value: "$header1": a valid HTTP header must consist of alphanumeric characters or '-' (e.g. 'X-Header-Name', regex used for validation is '[-A-Za-z0-9]+') 
            but expected 
            annotations.nginx.org/proxy-set-headers: Invalid value: "$header1": a valid HTTP header must consist of alphanumeric characters or '-' (e.g. 'X-Header-Name', regex used for validation is '[-A-Za-z0-9]+') 
            for the case of invalid nginx.org/proxy-set-headers annotation, single-value containing '$'
    --- FAIL: TestValidateNginxIngressAnnotations/invalid_nginx.org/proxy-set-headers_annotation,_single-value_containing_'{' 
 ```

Comment (pdabelf5) [reply internal/k8s/validation_test.go:1271]: Sticking with the current approach as the results are easier to understand if the const changes
